src/dtscalibration/io/apsensing.py

```diff
@@ -411,3 +423,263 @@ def metakey(meta, dict_to_parse, prefix):
     doc = doc_["WITSMLComposite"]["wellSet"]["well"]["wellboreSet"]
 
     return metakey({}, doc, ""), skip_chars
+
+
+def find_corresponding_tra_file(timestamp, tra_filelist):
+    """Finds match between timstamp extracted from .xml file and filepathlist containing tra_files.
+
+    Parameters
+    ----------
+    timestamp : str
+    tra_filelist : list of str
+
+    Returns:
+    --------
+    match_exists : boolean
+    file : str
+    """
+    match_exists = False
+    for _, file in enumerate(tra_filelist):
+        tra_timestamp = file[-18:-5]
```

Comment:
A shame that they don't add the actual timestamp to the .tra file... Would make this more robust and less ugly. But I don't see a good way to avoid this.

Maybe one magical number less this way:
```suggestion
        tra_timestamp = file.replace(".tra","")[-13:]
```

Reply:
Also, not looping through all files might be faster. This is functional, but for every .xml file you have to look at between 1 and _all_ .tra files.

```py
tra_files = "\n".join(tra_filelist) # make it one big string
tra_timestamps = re.findall(r"(\d{14}).tra", data)  # find 14 digits followed by .tra
if timestamp in tra_timestamps:
    return True
return False
```

The regex is explained here: https://regex101.com/r/B9al6Y/1

And then you can also remove the outer loop:
```py
tra_files = "\n".join(tra_filelist) # make it one big string
tra_timestamps = set(re.findall(r"(\d{14}).tra", data))  # find 14 digits followed by .tra

xml_timestamps = "\n".join(filepathlist)
xml_timestamps = set(re.findall(r"(\d{14}).xml", data)) # note that these are sets now

diff = xml_timestamps - tra_timestamps
if len(diff) > 0:
    msg = f"Not all .xml files have a matching .tra file.\n Missing are time following timestamps {diff}."
    raise ValueError(msg)

diff = tra_timestamps - xml_timestamps
if len(diff) > 0:
    msg = f"Not all .tra files have a matching .xml file.\n Missing are time following timestamps {diff}."
    raise ValueError(msg)
```